### PR TITLE
bugfix/20459-wrong-tooltip-labelHeight-with-className

### LINKED ISSUE #20459: Wrong tooltip labelHeight when styling with dedicated className
#### Expected behaviour
Value of labelHeight in positioner function should return correct tooltip height.

#### Actual behaviour
When we try to add bigger padding to span inside tooltip, but styling via custom className, some elements have wrong value of labelHeight inside positioner function.

This is super corner case and I doubt that anyone will have problem like that. This issue started occuring in app after update highcharts(10.0.0 -> 11.3.0), node and angular. Simple workaround (or just correct approach) is to style via .highcharts-tooltip class.

#### Live demo with steps to reproduce
https://jsfiddle.net/AfterlifeTw/zkqhLgf0/68/

just hover over points and check console for height value.

#### Product version
Highcharts 11.3.0

#### Affected browser(s)
Chrome

### PR DESCRIPTION
**Pull request by jedrzejruta**
Fixed #20459, custom `className` wasn't applied to tooltip's label properly, which disallowed styling elements via the custom class.

### CODE REVIEW

**Review by TorsteinHonsi — COMMENTED** [review]
Before we land on this solution, I would like know why the `tooltip.getClassName()` function doesn't cut it. 

The `className` option is already applied here

https://github.com/highcharts/highcharts/blob/1a8c96198a7ee09b3aeae4ff0efbd521cff69152/ts/Core/Tooltip.ts#L1119-L1120

So why doesn't it work in this case? Is it not propagated to the div?

**Comment by jedrzejruta** [comment]
Strangely, it does propagate the `className` into the div and SVG group (at least by inspecting DOM elements), but the tooltip label `className` property stays the same, and thus results in incorrect height apllied.

**Comment by jedrzejruta** [comment]
@TorsteinHonsi any thoughts on this PR?

**Comment by TorsteinHonsi** [comment]
@jedrzejruta I get the feeling we're barking up the wrong tree here... If you take a look at this fiddle using the latest GA release: https://jsfiddle.net/highcharts/un9zrt8a/

After 2000 milliseconds the class name is updated. And as you see, it actually does take effect (green background), but it looks like the label height is not always correct. What I can see is that the three last columns' tooltip display correctly, but the first one looks like this: 
Skjermbilde 2024-05-27 kl  13 00 52

Firefox and Chrome are the same in this concern.

Also, it looks like your test doesn't pick up the updated class name. I don't think `SVGElement.className` is kept in sync after updating. Instead, this works for me:

```js
        assert.ok(
            chart.tooltip.label.hasClass('custom'),
            'Provided className should be added to the tooltip label, #20459.'
        );
```

**Comment by jedrzejruta** [comment]
@TorsteinHonsi You are correct, the problem occurs only on initial render of the tooltip, and since the BBox of the label is aquired before the `className` is added, the resulting height stays incorrect. After the initial load, the BBox for the point's tooltip label is cached and not updated, so the tooltip height for the first hovered point isn't updated, even though the `className` is properly added to the DOM.

File: ts/Core/Tooltip.ts [hunk 1]
```diff
@@ -1068,15 +1068,16 @@ class Tooltip {
                         });
                     }
 
+                    // Set the stroke color of the box to reflect the point
```
Comment (TorsteinHonsi) [inline ts/Core/Tooltip.ts:1071]: This comment seems out of place. It's probably just a remnant of older code, so we can just remove it now.

File: ts/Core/Tooltip.ts [hunk 2]
```diff
@@ -1068,15 +1068,16 @@ class Tooltip {
...
+                    // Add class before the label BBox is calculated (#21035)
+                    label.addClass(tooltip.getClassName(point), true);
```
Comment (TorsteinHonsi) [inline ts/Core/Tooltip.ts:1073]: The `addClass` function is chainable, so we can save some bytes by doing

`label.addClass()
    .attr();`

Alternatively, I think we can just pass the class as an attribute. Though I am not sure if the required order is preserved. Can you check if this works?

```ts
                    label.attr({
                        // Add class before the label BBox is calculated (#21035)
                        'class': tooltip.getClassName(point),
                        text: text && (text as any).join ?
                            (text as any).join('') :
                            text
                    });
```

**Review by TorsteinHonsi — CHANGES_REQUESTED** [review]
Thanks, this is a nice and tight solution! Some inline comments to possible micro optimization.

File: ts/Core/Tooltip.ts (hunk 2 above)
Comment (jedrzejruta) [reply ts/Core/Tooltip.ts:1073]: Passing class as an attribute does the work and the order is preserved correctly, so I'd opt to go with this approach.

**Review by TorsteinHonsi — APPROVED** [review]

**Pull request merged**